dipy/reconst/ivim.py

```diff
@@ -0,0 +1,330 @@
+    ----------
+    data : array ([X, Y, Z, ...], g)
+        Data or response variables holding the data. Note that the last
+        dimension should contain the data. It makes no copies of data.
+
+    jac : bool
+        Use the Jacobian? Default: False
+
+    Returns
+    -------
+    ivim_params: the S0, f, D_star, D value for each voxel.
+
+    """
+    flat_data = data.reshape((-1, data.shape[-1]))
+    flat_x0 = x0.reshape(-1, x0.shape[-1])
+    # Flatten for the iteration over voxels:
+    bvals = gtab.bvals
+    ivim_params = np.empty((flat_data.shape[0], 4))
+
+    if routine == 'minimize':
+        result = _minimize(flat_data, bvals, flat_x0, ivim_params,
+                           bounds, tol, jac, algorithm, gtol, ftol, eps)
+    elif routine == "leastsq":
+        result = _leastsq(flat_data, bvals, flat_x0, ivim_params)
+    ivim_params.shape = data.shape[:-1] + (4,)
+    return ivim_params
+
+
+def _minimize(flat_data, bvals, flat_x0, ivim_params,
+              bounds, tol, jac, algorithm, gtol, ftol, eps):
+    """Use minimize for finding ivim_params"""
+    sum_sq = lambda params, bvals, signal: np.sum(
+        _ivim_error(params, bvals, signal)**2)
+
+    result = []
+    for vox in range(flat_data.shape[0]):
+        res = Optimizer(sum_sq,
+                        flat_x0[vox],
+                        args=(bvals, flat_data[vox]), bounds=bounds,
+                        tol=tol, method=algorithm, jac=jac,
+                        options={'gtol': gtol, 'ftol': ftol, 'eps': eps})
+        ivim_params[vox, :4] = res.xopt
+        result += [res]
+    return result
+
+
+def _leastsq(flat_data, bvals, flat_x0, ivim_params):
+    """Use minimize for finding ivim_params"""
+    result = []
+    for vox in range(flat_data.shape[0]):
+        res = leastsq(_ivim_error,
+                      flat_x0[vox],
+                      args=(bvals, flat_data[vox]))
+        ivim_params[vox, :4] = res[0]
+        result += [res]
+    return result
+
+
+def two_stage(data, gtab, x0,
+              split_b, jac, bounds, tol,
+              routine, algorithm, gtol, ftol, eps):
+    """
+    Fit the ivim params using a two stage fit
+
+    Parameters
+    ----------
+    data : array ([X, Y, Z, ...], g)
+        Data or response variables holding the data. Note that the last
+        dimension should contain the data. It makes no copies of data.
+
+    jac : bool
+        Use the Jacobian? Default: False
+
+    Returns
+    -------
+    ivim_params: the S0, f, D_star, D value for each voxel.
+
+    """
+    flat_data = data.reshape((-1, data.shape[-1]))
+    # Flatten for the iteration over voxels:
+    bvals_ge_split = gtab.bvals[gtab.bvals > split_b]
+    bvecs_ge_split = gtab.bvecs[gtab.bvals > split_b]
+    gtab_ge_split = gradient_table(bvals_ge_split, bvecs_ge_split.T)
+
+    tensor_model = TensorModel(gtab_ge_split)
+    tenfit = tensor_model.fit(data[..., gtab.bvals > split_b])
+
+    D_guess = mean_diffusivity(tenfit.evals)
```

Comment:
Here's a question. Is the mean diffusivity correct or should it be the norm? I'm thinking maybe the norm is the correct choice because it's a truer representation of the diffusion magnitude.